presto-hive/src/main/java/com/facebook/presto/hive/CachingFileOpener.java

```diff
@@ -0,0 +1,41 @@
+package com.facebook.presto.hive;
+
+import com.facebook.presto.cache.CacheManager;
+import com.facebook.presto.cache.CachingFileSystem;
+import org.apache.hadoop.fs.FSDataInputStream;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+
+import javax.inject.Inject;
+
+import java.io.IOException;
+import java.util.Optional;
+
+public class CachingFileOpener
+        implements FileOpener
+{
+    private CacheManager cacheManager;
+    private FileOpener fileOpener;
+
+    @Inject
+    public CachingFileOpener(
+            CacheManager cacheManager,
+            @ForWarmStorageAwareFileOpener FileOpener fileOpener)
+    {
+        this.cacheManager = cacheManager;
+        this.fileOpener = fileOpener;
+    }
+
+    @Override
+    public FSDataInputStream open(FileSystem fileSystem, Path path, Optional<byte[]> extraFileInfo)
+            throws IOException
+    {
+        if (fileSystem instanceof CachingFileSystem) {
+            CachingFileSystem cachingFileSystem = (CachingFileSystem) fileSystem;
+            return fileOpener.open(cachingFileSystem.getDataTier(), path, extraFileInfo);
+        }
+        else {
```

Comment:
redundant; also, instead of using `fileSystem instanceof CachingFileSystem`, just `checkArgument`

Reply:
Not sure why it is redundant? Its unlikely, but Its just a safety check. Anyways I will switch it to checkArgument.

Reply:
I mean `else {` is redundant. Just directly throw given the previous branch has already returned.